tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py

```diff
@@ -0,0 +1,160 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with shared weights between wrappers using a model with mevo kernel. """
+
+from copy import deepcopy
+
+import pytest
+import torch
+from torch import nn
+import torch.multiprocessing as mp
+from torch.optim import SGD
+
+from fairscale.experimental.nn import MEVO
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.utils.testing import dist_init, objects_are_equal, skip_if_single_gpu, teardown, temp_files_ctx
+
+VOCAB = 4
+D_MODEL = 2
+BS = 2
+SEQ = 3
+TILE = 2
+
+_large = True
+
+if _large:
+    VOCAB = 1024 * 50
+    D_MODEL = 1024
+    BS = 2
+    SEQ = 16
+    TILE = 16
+
+
+class Model(nn.Module):
+    def __init__(self, with_fsdp=False, wrap_middle="none"):
+        super().__init__()
+        self.l0 = nn.Embedding(VOCAB, D_MODEL).cuda().half()
+        nn.init.uniform_(self.l0.weight, -1.0e-1, 1.0e-1)
+        self.l1 = MEVO(self.l0.weight, tile_factor=TILE, reduction="sum")
+        self.middle = nn.Linear(D_MODEL, D_MODEL).cuda().half()
+        # LNs are not strictly needed for this test, but they help reduce the loss quickly
+        # and improves the numerical stability.
+        self.ln1 = nn.LayerNorm(D_MODEL).cuda().half()
+        self.ln2 = nn.LayerNorm(D_MODEL).cuda().half()
+
+        if with_fsdp:
+            # Shared layers much be un-flatten.
+            self.l0 = FSDP(self.l0, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1 = FSDP(self.l1, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1.append_shared_param(self.l0.module.weight)
```

Comment:
MEVO should also work for non-shared layers right? Curious what the reason was for having that support.

Reply:
The floor memory consumption won't be as good otherwise. If mevo is used without this sharding support, then its param and grad memory will stay unsharded (i.e. full) for the entire forward/backward pass. This reduces the memory usable by other layers. With this support, we shard them after the layer is done computing, leaving much more memory to other layers, like the attention layers.